metrics/ci_integrations/test/client/buildkite/test_utils/mock/buildkite_mock_server.dart

```diff
@@ -0,0 +1,241 @@
+
+    await request.response.close();
+  }
+
+  /// Returns the [BuildkiteBuildState], based on the `state` query parameter
+  /// of the given [request].
+  BuildkiteBuildState _extractBuildState(HttpRequest request) {
+    final state = request.uri.queryParameters['state'];
+
+    return const BuildkiteBuildStateMapper().map(state);
+  }
+
+  /// Returns the `per_page` query parameter of the given [request].
+  ///
+  /// Returns `null` if the `perPage` is `null`.
+  int _extractPerPage(HttpRequest request) {
+    final perPage = request.uri.queryParameters['per_page'];
+
+    if (perPage == null) return null;
+
+    return int.tryParse(perPage);
+  }
+
+  /// Returns the `page` query parameter of the given [request].
+  ///
+  /// Returns `null` if the `page` is `null`.
+  int _extractPage(HttpRequest request) {
+    final page = request.uri.queryParameters['page'];
+
+    if (page == null) return null;
+
+    return int.tryParse(page);
+  }
+
+  /// Generates a list of [BuildkiteBuild]s with the given [state].
+  ///
+  /// If the given [state] is null, the [BuildkiteBuildState.finished] is used.
+  List<BuildkiteBuild> _generateBuilds(BuildkiteBuildState state) {
+    return List.generate(
+      100,
+      (index) => BuildkiteBuild(
+        id: 'id',
+        number: index,
+        blocked: false,
+        state: state ?? BuildkiteBuildState.finished,
+        webUrl: 'url',
+        startedAt: DateTime(2020),
+        finishedAt: DateTime(2021),
+      ),
+    );
+  }
+
+  /// Generates a list of [BuildKiteArtifact]s.
+  List<BuildkiteArtifact> _generateArtifacts() {
+    return List.generate(
+      100,
+      (index) => BuildkiteArtifact(
+        id: 'id',
+        filename: 'coverage$index',
+        downloadUrl: 'url',
+        mimeType: 'json',
+      ),
+    );
+  }
+
+  /// Sets the next page url header using the given [request], [itemsCount],
+  /// [runsPerPage] and [pageNumber].
+  void _setNextPageUrlHeader(
+    HttpRequest request,
+    int itemsCount,
+    int runsPerPage,
+    int pageNumber,
+  ) {
+    final lastPageNumber = _getLastPageNumber(itemsCount, runsPerPage);
+
+    if (pageNumber >= lastPageNumber) return;
+
+    final requestUrl = request.requestedUri.toString();
+    final indexOfPageParam = requestUrl.indexOf("&page=");
+    final nextPageUrl = requestUrl.replaceRange(
+      indexOfPageParam,
+      requestUrl.length,
+      "&page=${pageNumber + 1}",
+    );
+
+    request.response.headers.set('link', '<$nextPageUrl> rel="next"');
+  }
+
+  /// Returns the last page number.
+  ///
+  /// Returns `1` if the given [perPage] or [total] parameter is `null`.
+  int _getLastPageNumber(int total, int perPage) {
+    if (perPage == null || total == null) return 1;
+
+    return max((total / perPage).ceil(), 1);
+  }
+
+  /// Chunks the given [items], limiting to the [limit],
+  /// starting from the [pageIndex].
+  List<T> _paginate<T>(List<T> items, [int limit, int pageIndex]) {
```

Comment:
Does it make sense to extract this method to some utility class to be able to reuse it in different mock servers? If so, the same for the `_writeResponse` method.